rfd/0000-sre-challenge.md

```diff
@@ -0,0 +1,213 @@
+
+### Caching, mTLS, and Delivery
+
+#### Caching
+
+The API will use a caching mechanism to store and retrieve the replica count of the Kubernetes deployment. This will reduce the number of API calls to the Kubernetes API server, improving performance and efficiency.
+
+##### Implementation Details:
+1. **In-Memory Cache**: Use an in-memory map to store the replica counts.
+2. **Synchronization**: Use a `sync.RWMutex` to manage concurrent access to the cache.
+3. **Kubernetes Informers**: Use Kubernetes informers to watch for changes to deployments and update the cache accordingly.
+   - **AddFunc**: Triggered when a deployment is added. It updates the cache.
+   - **UpdateFunc**: Triggered when a deployment is updated. It updates the cache.
+   - **DeleteFunc**: Triggered when a deployment is deleted. It removes the entry from the cache.
+
+#### mTLS
+
+The API will use mutual TLS (mTLS) for secure communication. This involves the following steps:
+- Generating a CA certificate and key.
+- Generating server and client certificates signed by the CA.
+- Configuring the server to require client certificates for authentication.
+- Configuring the client to use its certificate and the CA certificate to verify the server.
+
+### Security
+
+The chosen cipher suites for mTLS will be:
+- TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+- TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+- TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+- TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+
+Additional security measures include:
+- Strict validation of client certificates.
+
+### Testing
+
+Testing will be a critical part of ensuring the robustness and security of the API. The following testing strategies will be implemented:
+
+1. **Unit Tests**: These tests will cover individual functions and components of the Go application, ensuring they work as expected in isolation.
+2. **Integration Tests**: These tests will validate the interactions between the API endpoints and the Kubernetes cluster. They will be run in the KIND cluster to simulate a real Kubernetes environment.
+3. **End-to-End Tests**: These tests will cover the full workflow from deploying the application, interacting with the API, and verifying the expected outcomes. mTLS configurations will be tested to ensure secure communication.
+4. **Happy Path Tests**: These tests will cover scenarios where the API behaves as expected:
+   - Successfully retrieving the replica count.
+   - Successfully setting the replica count.
+5. **Unhappy Path Tests**: These tests will cover scenarios where the API encounters errors:
+   - Attempting to set an invalid replica count.
+   - Failing to retrieve the replica count due to missing deployment.
+
+### Delivery
+
+The API will be delivered as a Docker image, which can be deployed to any Kubernetes cluster using Helm charts. This ensures that the API can be easily deployed and managed in different environments.
```

Comment:
What kind of kubernetes resources are you planning to deploy with the helm chart?

Reply:
The Helm chart will deploy the following Kubernetes resources:

1. Deployment: This will manage the deployment of the API Docker container, ensuring the specified number of replicas are running.
2. Service: This will expose the API deployment within the Kubernetes cluster, allowing it to be accessed by other services or through port forwarding.
3. ClusterRole and ClusterRoleBinding: These will grant cluster-wide permissions necessary for the API to interact with Kubernetes resources.
4. Role and RoleBinding: These will grant namespace-specific permissions for the API to interact with Kubernetes resources.